core/src/main/java/io/grpc/Metadata.java

```diff
@@ -132,87 +135,145 @@ public Integer parseAsciiString(String serialized) {
     }
   };
 
-  /** All value lists can be added to. No value list may be empty. */
-  // Use LinkedHashMap for consistent ordering for tests.
-  private final Map<String, List<MetadataEntry>> store =
-      new LinkedHashMap<String, List<MetadataEntry>>();
-
-  /** The number of headers stored by this metadata.  */
-  private int storeCount;
+  /**
+   * Constructor called by the transport layer when it receives binary metadata.
+   */
+  Metadata(byte[]... binaryValues) {
+    this(binaryValues.length / 2, binaryValues);
+  }
 
   /**
    * Constructor called by the transport layer when it receives binary metadata.
    */
   // TODO(louiscryan): Convert to use ByteString so we can cache transformations
-  @Internal
-  public Metadata(byte[]... binaryValues) {
-    checkArgument(binaryValues.length % 2 == 0,
-        "Odd number of key-value pairs: %s", binaryValues.length);
-    for (int i = 0; i < binaryValues.length; i += 2) {
-      // The transport might provide an array with null values at the end.
-      if (binaryValues[i] == null) {
-        break;
-      }
-      String name = new String(binaryValues[i], US_ASCII);
-      storeAdd(name, new MetadataEntry(name.endsWith(BINARY_HEADER_SUFFIX), binaryValues[i + 1]));
-    }
+  Metadata(int usedNames, byte[]... binaryValues) {
+    // TODO(carl-mastrangelo): maybe remove this, unless
+    // https://bugs.openjdk.java.net/browse/JDK-6445664 is fixed.
```

Comment:
I don't really know how it works. There is also a difference between C1 and C2. There is probably a difference between hardware platforms and operating systems.

What I believe to be somewhat true for C2 on Linux x64, from the top of my head:

- Methods less than or equal to 35 bytecodes are inlined.
- Hot methods, determined by some measure of invocation count per unit of time, are inlined if less than or equal to 325 bytecodes. Rule of thumb is ~10k invocations in a short period (I guess millis).
- There is a max inlining depth. I think it's 9 by default. That is, the compiler will only inline nested method calls up to 9 levels deep.
- There is some limit of maximum code size. That is, even if all of the above is true, if the resulting method is larger than that limit, compilation and inlining will fail. Now I believe that's the size of the compiled code,  not bytecodes. I am not sure though.
- In order for inlining to work, a method needs to be devirtualized successfully. Static and final methods or methods that via class hierachy analysis were determined for a given call side to only be invoked on up to two different types. I think there is some trouble with interfaces, as they are not included in CHA. Don't recall what exactly the issue was.
- There is some maximum compilation time i.e. if a method takes longer than N millis to compile, compilation (and inlining) will fail. Not sure how the JIT recovers from that. Probably by trying to inline less aggressively.

Generally I wouldn't worry too much about it. Can check inlining logs of benchmarks from time to time, and see if there are any big offenders in our code.